### PR TITLE
Allow for structured variables on the scheduled product page

### PR DESCRIPTION
**Pull request by r-richardson**
This PR fixes the scheduled product page loading issues, which occur when job settings contain arrays, mentioned in https://progress.opensuse.org/issues/166154

example:
Screenshot from 2024-10-09 17-14-31

### CODE REVIEW

**Comment by perlpunk** [comment]
I found a bug. The display of multiline fields is now broken:
scheduled-product-master
scheduled-product-PR

**Review by okurz — CHANGES_REQUESTED** [review]
Please fix your commit that still says "WIP"

**Comment by Martchus** [comment]
And could you provide a screenshot so we can see how it looks like when one has such "structured variables" on a product?

File: assets/javascripts/openqa.js [hunk 1]
```diff
@@ -577,19 +577,43 @@ function renderComments(row) {
...
+    return document.createTextNode('');
   }
-  span.appendChild(document.createTextNode(value));
-  return span;
+  if (Array.isArray(value)) {
+    const fragment = document.createDocumentFragment();
+    value.forEach((item, index) => {
+      fragment.appendChild(renderHttpUrlAsLink(item));
+      if (index < value.length - 1) {
+        fragment.appendChild(document.createTextNode(', ')); // rator between items
```
Comment (okurz) [inline assets/javascripts/openqa.js:588]: what's a "rator"?

**Review by okurz — CHANGES_REQUESTED** [review]
What you put into the description of the pull request could also be in the commit message details. I don't know how you created your pull request but your git commit messages has *only* the subject line while the pull request description has more details. Please keep in mind that the github pull request description is only visible on github, the commit can be considered permanent information storage.

I can recommend the tool [hub](https://github.com/github/hub) (zypper in rubygem-hub) for easier PR creation. Also, I myself use a script [git-pr-last](http://github.com/okurz/scripts/tree/master/git-pr-last) to create a PR with proper description for these simple one-commit PRs:
```bash
$ cat $(which git-pr-last )
#!/bin/sh -e
target="${target:-"$USER"}"
git push $target && git show --no-patch --format=%B | hub pull-request -F -
```

You might want to use the more up-to-date tool https://github.com/cli/cli

See https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/CONTRIBUTING.md#coding-style for more details

File: assets/javascripts/openqa.js (hunk 1 above)
Comment (r-richardson) [reply assets/javascripts/openqa.js]: Added `sepe` :sweat_smile: :+1: 
And yes makes sense that the git message has higher prio > PR message, i've improved the commit message and will keep that in mind. Also i will have a look at hub, thanks :slightly_smiling_face:

**Review by okurz — APPROVED** [review]

File: assets/javascripts/openqa.js [hunk 2]
```diff
@@ -577,19 +577,43 @@ function renderComments(row) {
...
+  let lastIndex = 0;
+  let match;
+  const fragment = document.createDocumentFragment();
+  while ((match = urlRegex.exec(value)) !== null) {
+    if (match.index > lastIndex) {
+      fragment.appendChild(document.createTextNode(value.substring(lastIndex, match.index)));
+    }
+    const a = document.createElement('a');
+    a.href = match[0];
+    a.textContent = match[0];
```
Comment (b10n1k) [inline assets/javascripts/openqa.js:606]: `a.textContent = a.href = match[0];`?

File: assets/javascripts/openqa.js [hunk 3]
```diff
@@ -577,19 +577,43 @@ function renderComments(row) {
...
+    fragment.appendChild(a);
+    lastIndex = urlRegex.lastIndex;
+  }
+  if (lastIndex < value.length) {
+    fragment.appendChild(document.createTextNode(value.substring(lastIndex)));
+  }
+  if (!fragment.hasChildNodes()) {
+    return document.createTextNode(value);
+  }
+  return fragment;
```
Comment (b10n1k) [inline assets/javascripts/openqa.js:616]: Not sure but this looks like `return fragment.hasChildNodes() ? fragment : document.createTextNode(value);`

**Review by b10n1k — CHANGES_REQUESTED** [review]
Minor changes requested. they are calisthenic touches and nothing to do with the logic. Just following some good practices. thank you man and let me know if you have objections

**Review by Martchus — APPROVED** [review]

File: assets/javascripts/openqa.js [hunk 4]
```diff
@@ -577,19 +577,39 @@ function renderComments(row) {
...
-    span.appendChild(document.createTextNode(value.substr(0, match.index)));
-    span.appendChild(link);
-    value = value.substr(match.index + url.length);
+  if (!value) {
+    return document.createTextNode('');
   }
-  span.appendChild(document.createTextNode(value));
-  return span;
+  if (Array.isArray(value)) {
+    const fragment = document.createDocumentFragment();
```
Comment (b10n1k) [inline assets/javascripts/openqa.js:584]: it seems you can move this outside the if and then you can delete https://github.com/os-autoinst/openQA/pull/5972/files#diff-fae8284b51d426d67a2d9bb8590c1f8e06151d909a34a7e67f36bdc2038907e9R599 as well.

**Review by b10n1k — CHANGES_REQUESTED** [review]
Sorry but I just noticed that I missed to press the comment button or something in the morning and one of my comments was missing.

File: assets/javascripts/openqa.js (hunk 4 above)
Comment (r-richardson) [reply assets/javascripts/openqa.js:584]: yes, good catch :+1:

**Review by b10n1k — APPROVED** [review]

**Pull request merged**